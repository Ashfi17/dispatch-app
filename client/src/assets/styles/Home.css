table {
  border-radius: 8px;
  border-collapse: collapse;
}
.table-card {
  margin: 20px 80px;
  width: 80%;
}
table#table-container tr:nth-child(even) {
  background-color: #eee;
}
table#table-container tr:nth-child(odd) {
  background-color: #fff;
}
table#table-container th {
  color: white;
}
th,
td {
  text-align: left;
  text-align: center;
  border-collapse: collapse;
  font-weight: 600;
}
th {
  font-size: 18px;
  padding: 22px;
  background-color: crimson;
}
td {
  color: #858486;
  font-size: 16px;
  padding: 15px;
}
/* .export-btn {
  float: right;
  margin: 20px 10px;
  width: 140px;
  padding: 12px;
  background-color: crimson;
  outline: none;
  border: none;
  border-radius: 4px;
  color: white;
  font-weight: 600;
  font-size: 14px;
  word-spacing: 1px;
  cursor: pointer;
  border: 1px solid white;
} */
.search-sort-container {
  background-color: crimson;
  height: 50px;
  width: 80%;
  display: flex;
  flex-direction: row;
  padding: 16px;
  justify-content: space-evenly;
}
.search-field {
  width: 180px;
  height: 10px;
  margin: 10px 0px;
  border-radius: 18px;
  border: none;
  outline: none;
  padding: 12px;
}

.heading-buttons {
  width: 140px;
  height: 40px;
  margin: 10px 70px;
  border-radius: 6px;
  padding: 0px 10px;
  border: none;
  outline: none;
  color: white;
  cursor: pointer;
  background: none;
  border: 1px solid white;
}

.message {
  text-align: center;
  color: crimson;
  font-size: 16px;
  font-weight: 600;
}
