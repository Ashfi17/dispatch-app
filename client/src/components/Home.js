import React, { Component } from "react";
import { getDispatchInfo, search, exportExcel } from "../actions/Dispatch";
import { connect } from "react-redux";
import "../assets/styles/Home.css";
import jwt_decode from "jwt-decode";
export class Home extends Component {
  state = {
    search: "",
    descendingOrder: false,
    message: "",
  };
  componentDidMount() {
    if (localStorage.getItem("token")) {
      var decode = jwt_decode(localStorage.getItem("token"));
      console.log(decode);
      //checking if the token is expired or not
      if (Date.now() > decode.exp * 1000) {
        //if token is expired push to login page
        this.props.history.push("/login", {
          message: "Token expired please login again",
        });
      } else this.props.getDispatchInfo({ order: "desc" });
    } else {
      this.props.history.push("/login", {
        message: "No token present, please login",
      });
    }
  }
  //function to handle text change in search
  OnTextChange = (event) => {
    var decode = jwt_decode(localStorage.getItem("token"));
    this.setState({
      [event.target.name]: event.target.value,
    });
    if (Date.now() > decode.exp * 1000) {
      this.props.history.push("/login", {
        message: "Token expired please login again",
      });
    } else this.searchSourceCode();
  };
  timeOutId = null;

  //this function executes the action function after the user stops typing
  searchSourceCode = () => {
    clearTimeout(this.timeOutId);
    this.timeOutId = setTimeout(() => {
      if (!this.state.search) {
        this.props.getDispatchInfo({ order: "desc" });
      } else {
        this.props.search({ search: this.state.search });
      }
    }, 1000);
  };

  //function to sort the table content by startdate either ascending or descenfing
  onSortByStartDateSelected = () => {
    let { descendingOrder } = this.state;
    var decode = jwt_decode(localStorage.getItem("token"));
    this.setState({ descendingOrder: !descendingOrder }, () => {
      if (descendingOrder === true) {
        //checking if the token is expired
        if (Date.now() > decode.exp * 1000) {
          this.props.history.push("/login", {
            message: "Token expired please login again",
          });
        } else this.props.getDispatchInfo({ order: "desc" });
      } else {
        if (Date.now() > decode.exp * 1000) {
          this.props.history.push("/login", {
            message: "Token expired please login again",
          });
        } else this.props.getDispatchInfo({ order: "asc" });
      }
    });
  };

  render() {
    return (
      <div>
        <div className="table-card">
          <div className="search-sort-container">
            <input
              id="search"
              className="search-field"
              name="search"
              placeholder="search by source code"
              value={this.state.search}
              onChange={(e) => this.OnTextChange(e)}
            />
            <button
              id="sort-btn"
              className="heading-buttons"
              onClick={this.onSortByStartDateSelected}
            >
              Sort By Start Date
            </button>
            <button
              id="export-button"
              className="heading-buttons"
              onClick={async () => {
                var decode = jwt_decode(localStorage.getItem("token"));
                //checking is the token is expired
                if (Date.now() > decode.exp * 1000) {
                  this.props.history.push("/login", {
                    message: "Token expired please login again",
                  });
                } else {
                  //this function creates the excel sheet
                  await this.props.exportExcel();
                  this.setState({ message: this.props.excelMessage });
                  clearTimeout(this.timeOutId);
                  this.timeOutId = setTimeout(() => {
                    this.setState({ message: "" });
                  }, 5000);
                }
              }}
            >
              Export To excel
            </button>
            <button
              className="heading-buttons"
              onClick={() => this.props.history.push("/import-new")}
            >
              Add New Detail
            </button>
          </div>
          <table id="table-container">
            <tr>
              <th>Delivery No.</th>
              <th>Source Code</th>
              <th>Destination Code</th>
              <th>Transporter Code</th>
              <th>Vehicle Number</th>
              <th>Start Date</th>
              <th>End Date</th>
            </tr>
            {this.props.dispatchDetails.map((info, index) => (
              <tr key={index}>
                <td>{info.delivery_number}</td>
                <td>{info.source_code}</td>
                <td>{info.destination_code}</td>
                <td>{info.transporter_code}</td>
                <td>{info.vehicle_number}</td>
                <td>{new Date(info.start_date).toISOString().slice(0, 10)}</td>
                <td>{new Date(info.end_date).toISOString().slice(0, 10)}</td>
              </tr>
            ))}
          </table>
          <p className="message">{this.state.message}</p>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  dispatchDetails: state.DispatchReducer.dispatchDetails,
  excelMessage: state.DispatchReducer.excelMessage,
});

export default connect(mapStateToProps, {
  getDispatchInfo,
  search,
  exportExcel,
})(Home);
