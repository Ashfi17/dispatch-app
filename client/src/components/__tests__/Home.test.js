import React from "react";
import { shallow } from "enzyme";
import { Home } from "../Home";

const getDispatchInfo = jest.fn();
const search = jest.fn();
const exportExcel = jest.fn();
const history = {
  push: jest.fn(),
  // location: { state: { message: "Token Expired please login again" } },
};

const dispatchDetails = [
  {
    delivery_number: 1,
    source_code: "S001",
    destination_code: "D001",
    transporter_code: "T001",
    vehicle_number: "MH04AB1234",
    start_date: "2020-01-21",
    end_date: "2020-02-22",
  },
];

const wrapper = shallow(
  <Home
    getDispatchInfo={getDispatchInfo}
    search={search}
    exportExcel={exportExcel}
    dispatchDetails={dispatchDetails}
    history={history}
  />
);

beforeAll(() => {
  localStorage.setItem("token", "abcdefgh");
});
afterAll(() => {
  localStorage.removeItem("token");
});

describe("test suite for Home component", () => {
  it("should mount the component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should test  the functions is working when user searches ", () => {
    const OnTextChange = jest.spyOn(wrapper.instance(), "OnTextChange");
    const searchSourceCode = jest.spyOn(wrapper.instance(), "searchSourceCode");
    const event = {
      target: "search",
    };
    wrapper.find("#search").simulate("change", event);
    expect(OnTextChange).toBeCalled();
    expect(searchSourceCode).toBeCalled();
  });
  it("should test the button clicks and check if the sortByDate function is being called", () => {
    const onSortByStartDateSelected = jest.spyOn(
      wrapper.instance(),
      "onSortByStartDateSelected"
    );
    expect(wrapper.find("#sort-btn").simulate("click"));
    wrapper.instance().forceUpdate();
    expect(onSortByStartDateSelected).toBeCalled();
  });
  it("should test the button clicks and check if the action functions are being called", () => {
    wrapper.find("#export-button").simulate("click");
    expect(exportExcel).toBeCalled();
  });
});
