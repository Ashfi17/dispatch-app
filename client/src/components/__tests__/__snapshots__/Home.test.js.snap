// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test suite for Home component should mount the component 1`] = `
<div>
  <div
    className="table-card"
  >
    <div
      className="search-sort-container"
    >
      <input
        className="search-field"
        id="search"
        name="search"
        onChange={[Function]}
        placeholder="search by source code"
        value=""
      />
      <button
        className="heading-buttons"
        id="sort-btn"
        onClick={[Function]}
      >
        Sort By Start Date
      </button>
      <button
        className="heading-buttons"
        id="export-button"
        onClick={[Function]}
      >
        Export To excel
      </button>
      <button
        className="heading-buttons"
        onClick={[Function]}
      >
        Add New Detail
      </button>
    </div>
    <table
      id="table-container"
    >
      <tr>
        <th>
          Delivery No.
        </th>
        <th>
          Source Code
        </th>
        <th>
          Destination Code
        </th>
        <th>
          Transporter Code
        </th>
        <th>
          Vehicle Number
        </th>
        <th>
          Start Date
        </th>
        <th>
          End Date
        </th>
      </tr>
      <tr
        key="0"
      >
        <td>
          1
        </td>
        <td>
          S001
        </td>
        <td>
          D001
        </td>
        <td>
          T001
        </td>
        <td>
          MH04AB1234
        </td>
        <td>
          21-01-2020
        </td>
        <td>
          22-02-2020
        </td>
      </tr>
    </table>
    <p
      className="message"
    />
  </div>
</div>
`;
